File: 200-number-of-islands/number-of-islands.cpp
class Solution {
public:
    void helper(int i, int j, vector<vector<char>>& grid, int n, int m)
    {
        if(i < 0 || j < 0 || i == n || j == m || grid[i][j] != '1' ) return;
        grid[i][j] = 2;
        helper(i+1, j, grid, n,m);
        helper(i, j+1, grid, n, m);
        helper(i-1, j , grid, n, m);
        helper(i, j-1, grid, n, m);

    }
    int numIslands(vector<vector<char>>& grid) {
        int n = grid.size();
        int m = grid[0].size();
        int count = 0;
        for(int i = 0; i<n ;i++)
        {
            for(int j = 0; j<m; j++)
            {
                if(grid[i][j] == '1')
                {
                    helper(i, j , grid, n, m);
                    count+=1;
                }
            }
        }
        return count;
    }
};

File: 200-number-of-islands/number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid:
            return 0 
        n, m = len(grid), len(grid[0])
        dirs = [[1, 0], [0,1], [-1, 0], [0,  -1]]
        queue = deque()
        visited = set()
        def bfs(row, col):
            nonlocal visited
            nonlocal queue
            queue.append((row, col))
            while queue:
                tR, tC = queue.popleft()
                visited.add((row, col))
                for x, y in dirs:
                    nR, nC = (x + tR), (y + tC)
                    if nR >= 0 and nC >=0 and  nR < n and nC < m and grid[nR][nC] == "1" and (nR, nC) not in visited:
                        queue.append((nR, nC))
                        visited.add((nR,nC))
        numIslands = 0
        for row in range(n):
            for col in range(m):
                if grid[row][col] == "1" and (row, col) not in visited:
                    bfs(row, col)
                    numIslands += 1
        return numIslands